src/parsimmon.js

```diff
@@ -9,6 +9,161 @@ function Parsimmon(action) {
 
 var _ = Parsimmon.prototype;
 
+function times(n, f) {
+  var i = 0;
+  for (i; i < n; i++) {
+    f(i);
+  }
+}
+
+function forEach(f, arr) {
+  times(arr.length, function(i) {
+    f(arr[i], i, arr);
+  });
+}
+
+function reduce(f, seed, arr) {
+  forEach(function(elem, i, arr) {
+    seed = f(seed, elem, i, arr);
+  }, arr);
+  return seed;
+}
+
+function map(f, arr) {
+  return reduce(
+    function(acc, elem, i, a) {
+      return acc.concat([f(elem, i, a)]);
+    },
+    [],
+    arr
+  );
+}
```

Comment:
If you don't mind, I think moving `arr` to be the first argument in these functions will make the code read better

Reply:
I wrote this to be more in the style of like ramda, where the point in the function comes last.  Putting the point at the beginning of the function makes it harder to work with if you start to use them in a partially applied way.

If you really like the array-first style, what do you think about doing a newtype wrapper style set of helpers?

like
```javascript
function Func(f) {
    this._f = f;
}

Func.prototype.times = function times(n) {
    var self = this;
    return new Func(function () {
        var i = 0;
        for (i; i < n; i++) {
            self._f(i);
        }
    });
};

Func.prototype.run = function run() {
    return this._f();
};

function Enum(arr) {
    this._arr = arr;
}

Enum.prototype.forEach = function forEach(f) {
    return new Enum(new Func(f).times(this._arr.length).run());
};

Enum.prototype.reduce = function reduce(f, seed) {
    this.forEach(function (elem, i, arr) {
        seed = f(seed, elem, i, arr);
    });
    return seed;
};

Enum.prototype.map = function map(f) {
    return new Enum(this.reduce(function (acc, elem, i, a) {
        return acc.concat([f(elem, i, a)]);
    }, []));
};

Enum.prototype.value = function value() {
    return this._arr;
};

new Enum([1, 2, 3]).map(function (x) { return x + 5; }).value();
//[ 5, 6, 7 ]
```

Reply:
I actually use to be a ramda 🐑λ  contributor so I definitely know the style :) anyway it's not really a big deal so if you really like it array-last just keep it 👌

Reply:
I think the newtype wrapper approach lends itself more towards the es6 built in array methods down the road.  But I personally use point free style heavily in my own code, but as you know that typically doesn't come up much unless you're doing composition, and I don't even have a compose/pipe helper currently.